src/serde_helpers.rs

```diff
@@ -0,0 +1,175 @@
+    serializer.serialize_i64(*val as i64)
+}
+
+/// Attempts to serialize a u64 as an i32. Errors if an exact conversion is not possible.
+pub fn serialize_u64_as_i32<S: Serializer>(val: &u64, serializer: S) -> Result<S::Ok, S::Error> {
+    match i32::try_from(*val) {
+        Ok(val) => serializer.serialize_i32(val),
+        Err(_) => Err(ser::Error::custom(format!("cannot convert {} to i32", val))),
+    }
+}
+
+/// Attempts to serialize a u64 as an i64. Errors if an exact conversion is not possible.
+pub fn serialize_u64_as_i64<S: Serializer>(val: &u64, serializer: S) -> Result<S::Ok, S::Error> {
+    match i64::try_from(*val) {
+        Ok(val) => serializer.serialize_i64(val),
+        Err(_) => Err(ser::Error::custom(format!("cannot convert {} to i64", val))),
+    }
+}
+
+/// Deserializes a chrono::DateTime<Utc> from the extended JSON representation of DateTime.
+pub fn deserialize_chrono_datetime_from_ext_json<'de, D>(
+    deserializer: D,
+) -> Result<DateTime<Utc>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let bson = Bson::deserialize(deserializer)?;
+    match bson {
+        Bson::DateTime(date) => Ok(date),
+        _ => Err(de::Error::custom(
+            "cannot convert extended JSON to DateTime",
+        )),
+    }
+}
+
+/// Deserializes a DateTime from the extended JSON representation.
+pub fn deserialize_bson_datetime_from_ext_json<'de, D>(
+    deserializer: D,
+) -> Result<crate::DateTime, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let bson = Bson::deserialize(deserializer)?;
+    match bson {
+        Bson::DateTime(date) => Ok(crate::DateTime(date)),
+        _ => Err(de::Error::custom(
+            "cannot convert extended JSON to DateTime",
+        )),
+    }
+}
+
+/// Deserializes an ISO string from a DateTime.
+pub fn deserialize_iso_string_from_datetime<'de, D>(deserializer: D) -> Result<String, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let date = crate::DateTime::deserialize(deserializer)?;
+    Ok(date.to_string())
+}
+
+/// Serializes an ISO string as a DateTime.
+pub fn serialize_iso_string_as_datetime<S: Serializer>(
+    val: &str,
+    serializer: S,
+) -> Result<S::Ok, S::Error> {
+    let date = DateTime::from_str(val)
+        .map_err(|_| ser::Error::custom(format!("cannot convert {} to DateTime", val)))?;
+    Bson::DateTime(date).serialize(serializer)
+}
+
+/// Deserializes an ObjectId from the extended JSON representation.
+pub fn deserialize_object_id_from_ext_json<'de, D>(deserializer: D) -> Result<ObjectId, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let bson = Bson::deserialize(deserializer)?;
+    match bson {
+        Bson::ObjectId(oid) => Ok(oid),
+        _ => Err(de::Error::custom(
+            "cannot convert extended JSON to ObjectId",
+        )),
+    }
+}
+
+/// Serializes a hex string as an ObjectId.
+pub fn serialize_hex_string_as_object_id<S: Serializer>(
```

Comment:
No corresponding deserialize function for this one since checking for a hex string is already default behavior.